### PR TITLE
IGNITE-23831 Sql. ExecutionService must create unique executionId for retries for same queryId

### PR DESCRIPTION
**Pull request by korlov42**
https://issues.apache.org/jira/browse/IGNITE-23831

-------------
Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/message/ErrorMessage.java [hunk 1]
```diff
@@ -34,6 +34,8 @@ public interface ErrorMessage extends NetworkMessage, Serializable {
      */
     UUID queryId();
 
+    int executionToken();
```
Comment (xtern) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/message/ErrorMessage.java:37]: I would recommend adding javadoc for consistency with other methods.

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImplTest.java [hunk 2]
```diff
@@ -1081,6 +1041,70 @@ public void timeoutFiredOnInitialization() throws Throwable {
...
+                            .thenCompose(c -> this.apply(c, remainingAttempts - 1));
+                }
+
+                return previousStep;
+            }
+        };
+
+        int retryCount = 20;
+        await(retryChainBuilder.apply(null, retryCount));
+
```
Comment (xtern) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImplTest.java:1102]:
```suggestion

        assertThat(exceptions, hasSize(retryCount));

```

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImplTest.java [hunk 3]
```diff
@@ -462,48 +464,6 @@ void testErrorOnCursorInitialization() throws InterruptedException {
         awaitContextCancellation(execNodes);
     }
 
-    /**
-     * The very simple case where a query is cancelled in the middle of a normal execution on non-initiator node.
-     */
-    @Test
-    public void testCancelOnRemote() throws InterruptedException {
```
Comment (xtern) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImplTest.java:469]: Just curious. Were we really able to cancel a request from any node?

**Review by xtern — APPROVED** [review]

**Review by zstan — APPROVED** [review]

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java [hunk 4]
```diff
@@ -600,13 +589,10 @@ public void onDisappeared(ClusterNode member) {
...
     /** Returns local fragments for the query with given id. */
     public List<AbstractNode<?>> localFragments(UUID queryId) {
-        DistributedQueryManager mgr = queryManagerMap.get(queryId);
-
-        if (mgr == null) {
-            return List.of();
-        }
-
-        return mgr.localFragments();
+        return queryManagerMap.entrySet().stream()
```
[the remark this answers is not in the document]
Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java:593]: done, thanks

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/message/ErrorMessage.java (hunk 1 above)
Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/message/ErrorMessage.java:40]: done, thanks

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImplTest.java (hunk 3 above)
Comment (korlov42) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImplTest.java:469]: technically, yep. We were cancelled any fragment, and the QueryCancellation error was propagated to head of the query, and then back to remaining fragments.

but it never was in use, and now we have another mechanism for cancellation

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImplTest.java [hunk 5]
```diff
@@ -1081,6 +1041,70 @@ public void timeoutFiredOnInitialization() throws Throwable {
         }
     }
 
+    @Test
+    void executionsWithTheSameQueryIdMustNotInterfere() {
+        QueryPlan plan = prepare("SELECT * FROM test_tbl", operationContext(null).build());
```
[the remark this answers is not in the document]
Comment (korlov42) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImplTest.java]: replaced, thanks

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImplTest.java [hunk 6]
```diff
@@ -1081,6 +1041,70 @@ public void timeoutFiredOnInitialization() throws Throwable {
...
+                        .build()
+                );
+            } else {
+                original.onMessage(nodeName, msg);
+            }
+
+            return nullCompletedFuture();
+        });
+
+        SqlOperationContext ctx = operationContext(null).build();
```
[the remark this answers is not in the document]
Comment (korlov42) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImplTest.java]: replaced, thanks

**Pull request merged**